File: app/src/main/java/fr/xgouchet/chronorg/provider/dao/BaseDao.java
package fr.xgouchet.chronorg.provider.dao;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.WorkerThread;

/**
 * @author Xavier Gouchet
 */
public abstract class BaseDao<T> {

    @NonNull private final SQLiteOpenHelper openHelper;
    @NonNull private final String tableName;

    public BaseDao(@NonNull SQLiteOpenHelper openHelper, @NonNull String tableName) {
        this.openHelper = openHelper;
        this.tableName = tableName;
    }

    @WorkerThread
    public long insert(ContentValues cv) {
        SQLiteDatabase db = openHelper.getWritableDatabase();

        long insertedId = -1;

        db.beginTransaction();
        try {
            insertedId = db.insert(tableName, null, cv);
            db.setTransactionSuccessful();
        } finally {
            db.endTransaction();
        }

        return insertedId;
    }

    public Cursor query(@Nullable String[] projection,
                        @Nullable String selection,
                        @Nullable String[] selectionArgs,
                        @Nullable String sortOrder) {
        SQLiteDatabase db = openHelper.getReadableDatabase();

        return db.query(tableName, projection, selection, selectionArgs, null, null, sortOrder);
    }


    public int delete(@Nullable String selection,
                      @Nullable String[] selectionArgs) {
        SQLiteDatabase db = openHelper.getWritableDatabase();

        int deleted = 0;

        db.beginTransaction();
        try {
            deleted = db.delete(tableName, selection, selectionArgs);
            db.setTransactionSuccessful();
        } finally {
            db.endTransaction();
        }

        return deleted;
    }

    public int update(@Nullable ContentValues values,
                      @Nullable String selection,
                      @Nullable String[] selectionArgs) {
        if (values == null) return 0;
        int updated = 0;

        SQLiteDatabase db = openHelper.getWritableDatabase();
        db.beginTransaction();
        try {
            updated = db.update(tableName, values, selection, selectionArgs);
            db.setTransactionSuccessful();
        } finally {
            db.endTransaction();
        }

        return updated;
    }
}

File: khronorg/src/main/java/fr/xgouchet/khronorg/provider/dao/BaseDao.kt
package fr.xgouchet.khronorg.provider.dao

import android.annotation.SuppressLint
import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteOpenHelper

/**
 * @author Xavier Gouchet
 */
class BaseDao(private val openHelper: SQLiteOpenHelper,
              private val tableName: String) {


    fun insert(cv: ContentValues): Long {
        val db = openHelper.writableDatabase

        var insertedId: Long

        db.beginTransaction()
        try {
            insertedId = db.insert(tableName, null, cv)
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }

        return insertedId
    }

    @SuppressLint("Recycle")
    fun query(projection: Array<String>?,
              selection: String?,
              selectionArgs: Array<String>?,
              sortOrder: String?): Cursor {
        val db = openHelper.readableDatabase

        return db.query(tableName, projection, selection, selectionArgs, null, null, sortOrder)
    }


    fun delete(selection: String?,
               selectionArgs: Array<String>?): Int {
        val db = openHelper.writableDatabase

        var deleted: Int

        db.beginTransaction()
        try {
            deleted = db.delete(tableName, selection, selectionArgs)
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }

        return deleted
    }

    fun update(values: ContentValues?,
               selection: String?,
               selectionArgs: Array<String>?): Int {
        if (values == null) return 0
        var updated: Int

        val db = openHelper.writableDatabase
        db.beginTransaction()
        try {
            updated = db.update(tableName, values, selection, selectionArgs)
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }

        return updated
    }
}
